libs/librepcb/common/attributes/attributelist.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * LibrePCB - Professional EDA for everyone!
+ * Copyright (C) 2013 LibrePCB Developers, see AUTHORS.md for contributors.
+ * http://librepcb.org/
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*****************************************************************************************
+ *  Includes
+ ****************************************************************************************/
+#include <QtCore>
+#include "attributelist.h"
+#include "../fileio/xmldomelement.h"
+
+/*****************************************************************************************
+ *  Namespace
+ ****************************************************************************************/
+namespace librepcb {
+
+/*****************************************************************************************
+ *  Constructors / Destructor
+ ****************************************************************************************/
+
+AttributeList::AttributeList() noexcept
+{
+}
+
+AttributeList::AttributeList(const AttributeList& other) noexcept
+{
+    foreach (const QSharedPointer<Attribute>& attr, other.mAttributes) {
+        mAttributes.append(QSharedPointer<Attribute>::create(*attr));
+    }
+}
+
+AttributeList::AttributeList(const XmlDomElement& domElement) throw (Exception)
+{
+    for (XmlDomElement* node = domElement.getFirstChild("attribute", false);
+         node; node = node->getNextSibling("attribute"))
+    {
+        QSharedPointer<Attribute> attr = QSharedPointer<Attribute>::create(*node); // can throw
+        if (value(attr->getKey())) {
+            throw RuntimeError(__FILE__, __LINE__, attr->getKey(),
+                QString(tr("The attribute \"%1\" exists multiple times in \"%2\"."))
+                .arg(attr->getKey(), domElement.getDocFilePath().toNative()));
+        }
+        mAttributes.append(attr);
+    }
+}
+
+AttributeList::~AttributeList() noexcept
+{
+}
```

Comment:
Just to be consistent over the whole project. You could see this as a design guideline ;)